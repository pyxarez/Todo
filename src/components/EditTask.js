import React from 'react';

import './EditTask.res/style.css';

EditTask.propTypes = {
  onSave: React.PropTypes.func.isRequired,
  onCancel: React.PropTypes.func.isRequired,
  children: React.PropTypes.oneOfType(
    [
      React.PropTypes.arrayOf(React.PropTypes.element),
      React.PropTypes.bool,
      React.PropTypes.object
    ]
  ),
}

export default function EditTask({ onSave, onCancel, children }) {
  return (
    <div className="my-editTask-component">
      <div className="buttons-container">
        <button onClick={onSave}>Save changes</button>
        <button onClick={onCancel}>Cancel</button>
      </div>
      {children}
    </div>
  );
}
