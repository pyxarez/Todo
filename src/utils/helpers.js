export function findTarget(target, storage) {
  for (var i = 0; i < storage.length; i++) {
    const storageElem = storage[i];

    if (storageElem.id === +target) {
      const returningBundle = {
        category: storageElem,
        storage,
        index: i
      }

      return returningBundle;
    } else if (storageElem.nested.length > 0) {
      const returningBundle = findTarget(target, storageElem.nested);
      if (Object.keys(returningBundle).length === 3) return returningBundle;
    }
  }

  return {};
}

export const validateInput = (value) =>  {
  return !!value.trim()
};

const containsOnlyNumber = (string) => {
  return string.match(/^\d+$/);
}

export const toNumberKeyValues = (object) =>  {
  let newObject = {};
  
  for (let key in object) {
    if ( object.hasOwnProperty(key) && object[key] ){
      newObject[key] = containsOnlyNumber(object[key])
        ? parseInt(object[key], 10)
        : object[key];
    }
  }

  return newObject;
};

export function deepClone(target) {
  return JSON.parse(JSON.stringify(target));
}
