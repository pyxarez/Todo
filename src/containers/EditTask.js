import React, { Component, PropTypes } from 'react';
import { browserHistory } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { validateInput } from '../utils/helpers';

import { saveTaskChanges } from '../store/actions/TaskListActions';

import EditTask from '../components/EditTask';

export class EditTaskContainer extends Component {
  static propTypes = {
    task: PropTypes.shape({
      id: PropTypes.node.isRequired,
      title: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
      description: PropTypes.string.isRequired,
    }),
    categoryId: PropTypes.number.isRequired,
    saveTaskChanges: PropTypes.func.isRequired,
  }

  onSave = () => {
    const title = this.titleInput.value;
    const isDone = this.doneCheckBox.checked;
    const description = this.taskDescription.value;

    if (!validateInput(title)) {
      alert('Enter correct task title');
      return;
    }

    const options = {
      categoryId: this.props.categoryId,
      id: this.props.task.id,
      title,
      isDone,
      description
    };

    this.props.saveTaskChanges(options);
    browserHistory.goBack();
  }

  handleCancelClick = () => {
    browserHistory.goBack();
  }

  render() {
    const {
      title,
      isDone,
      description
    } = this.props.task;

    return (
      <EditTask onSave={this.onSave} onCancel={this.handleCancelClick}>
        <div>
          <input ref={(titleInput) => {this.titleInput = titleInput}} type="text" defaultValue={title}/>
          <input ref={(doneCheckBox) => {this.doneCheckBox = doneCheckBox}} type="checkbox" defaultChecked={isDone}/>
        </div>
        <textarea
          ref={(taskDescription) => {this.taskDescription = taskDescription}}
          defaultValue={description}
          className="description"
          name="description"
          cols="30"
          rows="30" />
      </EditTask>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    saveTaskChanges: bindActionCreators(saveTaskChanges, dispatch),
  }
}

export default connect(null, mapDispatchToProps)(EditTaskContainer)