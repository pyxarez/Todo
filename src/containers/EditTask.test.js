import React from 'react';
import { mount, shallow } from 'enzyme';

import { EditTaskContainer } from './EditTask';
import EditTask from '../components/EditTask';


jest.mock('../utils/helpers', () => {
  let returningValue = true;
  
  const gettingTrue = () => { returningValue = true };
  const gettingFalse = () => { returningValue = false };
  const validateInput = () => returningValue;
  
  return {
    gettingTrue,
    gettingFalse,
    validateInput
  };
});

jest.mock('react-router', () => {
  return {
    browserHistory: {
      goBack: jest.fn()
    }
  };
});


import { browserHistory } from 'react-router';
import helpers from '../utils/helpers';

const setup = () => {
 const props = {
    task: {
      id: 0,
      title: 'Task title',
      isDone: false,
      description: 'Lol'
    },
    categoryId: 0,
    saveTaskChanges: jest.fn()
    
  };

  const wrapper = mount(<EditTaskContainer {...props}/>);

  return {
    wrapper,
    props
  }
};

describe('Containers :: EditTaskContainer', () => {
  it('should render self and subcomponents', () => {
    const { props } = setup();
    const wrapper = shallow(<EditTaskContainer {...props} />);
    
    expect(wrapper).toMatchSnapshot();
  });

  it('should call saveTaskChanges and browserHistory.goBack', () => {
    const { wrapper, props } = setup();
    expect(wrapper).toMatchSnapshot();
    
    const editTaskComponent = wrapper.find(EditTask);
    
    const titleInput = wrapper.find('input').at(0);
    const isDoneInput = wrapper.find('input').at(1);
    const descriptionArea = wrapper.find('textarea');
    
    const title = 'New title';
    const isDone = true;
    const description = '1 2 3 4 5 6 7 8';
    
    titleInput.node.value = title;
    isDoneInput.node.checked = isDone;
    descriptionArea.node.value = description;

    editTaskComponent.prop('onSave')();
    
    expect(props.saveTaskChanges).toBeCalledWith({
      categoryId: props.categoryId,
      id: props.task.id,
      title,
      isDone,
      description
    });
    expect(browserHistory.goBack).toBeCalled();
  });

  it('should call browserHistory.goBack on cancel click', () => {
    const { wrapper } = setup();

    wrapper.find(EditTask).prop('onCancel')();
    expect(browserHistory.goBack).toBeCalled();
  });

  it('should call alert due to the invalid input value', () => {
    helpers.gettingFalse();
    const { wrapper, props } = setup();

    window.alert = jest.fn();
    const editTaskComponent = wrapper.find(EditTask);
    
    editTaskComponent.prop('onSave')();
    
    expect(window.alert)
      .toBeCalledWith('Enter correct task title');
  });
});
