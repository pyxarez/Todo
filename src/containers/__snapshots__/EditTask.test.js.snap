exports[`Containers :: EditTaskContainer should call saveTaskChanges and browserHistory.goBack 1`] = `
<EditTaskContainer
  categoryId={0}
  saveTaskChanges={[Function]}
  task={
    Object {
      "description": "Lol",
      "id": 0,
      "isDone": false,
      "title": "Task title",
    }
  }>
  <EditTask
    onCancel={[Function]}
    onSave={[Function]}>
    <div
      className="my-editTask-component">
      <div
        className="buttons-container">
        <button
          onClick={[Function]}>
          Save changes
        </button>
        <button
          onClick={[Function]}>
          Cancel
        </button>
      </div>
      <div>
        <input
          defaultValue="Task title"
          type="text" />
        <input
          defaultChecked={false}
          type="checkbox" />
      </div>
      <textarea
        className="description"
        cols="30"
        defaultValue="Lol"
        name="description"
        rows="30" />
    </div>
  </EditTask>
</EditTaskContainer>
`;

exports[`Containers :: EditTaskContainer should render self and subcomponents 1`] = `
<EditTask
  onCancel={[Function]}
  onSave={[Function]}>
  <div>
    <input
      defaultValue="Task title"
      type="text" />
    <input
      defaultChecked={false}
      type="checkbox" />
  </div>
  <textarea
    className="description"
    cols="30"
    defaultValue="Lol"
    name="description"
    rows="30" />
</EditTask>
`;
